src/psyclone/domain/lfric/metadata_to_arguments_rules.py

```diff
@@ -0,0 +1,614 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2023, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
+
+'''This module implements a class that encapsulates rules that map
+   LFRic kernel metadata to kernel arguments.
+
+'''
+from psyclone.domain.lfric import LFRicConstants
+from psyclone.domain.lfric.kernel import (
+    OperatorArgMetadata, ColumnwiseOperatorArgMetadata, FieldArgMetadata,
+    FieldVectorArgMetadata, InterGridArgMetadata, InterGridVectorArgMetadata,
+    ScalarArgMetadata)
+from psyclone.errors import InternalError
+
+# pylint: disable=too-many-branches
```

Comment:
I don't want the methods to be abstract as I want to be able to only override certain methods. An example of this is the `arg_index_to_metadata_index.py` subclass where we only override what we need to. Yes the 3 pylint directives are required in order to get rid of the pylint warnings - but I'm not sure that is what you were asking?

Reply:
The `too-many-branches` and `too-many-statements` are actually for the `_generate` method so I've moved them to there. This is a necessarily quite a complex method but I guess I could look to split it. The `too-few-public-methods` is because we have designed the class this way (as requested!) i.e. with a stateless interface, so there is only one way to use it `arguments = MetadataToArgumentsRules.mapping(metadata)`.

Reply:
OK. Thanks for pushing the two down to `_generate`.